libcaf_core/caf/meta/handler.hpp

```diff
@@ -0,0 +1,114 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/main/LICENSE.
+
+#pragma once
+
+#include "caf/detail/core_export.hpp"
+#include "caf/fwd.hpp"
+#include "caf/type_id_list.hpp"
+#include "caf/type_list.hpp"
+
+#include <string>
+
+namespace caf::meta {
+
+/// Descriptor for a message handler handler.
+struct handler {
+  type_id_list inputs;
+  type_id_list outputs;
+};
+
+/// @relates handler
+inline bool operator==(const handler& lhs, const handler& rhs) noexcept {
+  return lhs.inputs == rhs.inputs && lhs.outputs == rhs.outputs;
+}
+
+/// @relates handler
+inline bool operator!=(const handler& lhs, const handler& rhs) noexcept {
+  return !(lhs == rhs);
+}
+
+/// @relates handler
+CAF_CORE_EXPORT std::string to_string(const handler& hdl);
+
+/// Represents a list of message handlers for describing the interface of a
+/// statically typed actor.
+struct CAF_CORE_EXPORT handler_list {
+  /// The number of handlers in this list.
+  size_t size;
```

Comment:
```suggestion
  const size_t size;
```
This could help the compiler optimize things.

Reply:
If we would make that `const`, we would very subtly break the copy constructor of `handler_list`. The core guidelines have this to say about `const` members: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-constref. I'd actually go as far as to say member variables should *never* be a reference or `const`. I know we do have a few places in CAF where we still hold references as member variables. Reminds me that we should clean those up.

Reply:
Forgot that part entirely :flushed: . Thanks for taking care of the details.